packages/runtime-core/src/componentProps.ts

```diff
@@ -19,24 +19,30 @@ import { Data, ComponentInternalInstance } from './component'
 
 export type ComponentPropsOptions<P = Data> =
   | ComponentObjectPropsOptions<P>
-  | string[]
+  | Extract<keyof P, string>[]
 
 export type ComponentObjectPropsOptions<P = Data> = {
   [K in keyof P]: Prop<P[K]> | null
 }
 
 export type Prop<T> = PropOptions<T> | PropType<T>
 
+type DefaultFactory<T> = () => T | null | undefined
+
 interface PropOptions<T = any> {
   type?: PropType<T> | true | null
   required?: boolean
-  default?: T | null | undefined | (() => T | null | undefined)
-  validator?(value: unknown): boolean
+  default?:
+    | DefaultFactory<T>
+    | null
+    | undefined
+    | (T extends Function ? never : T) // If type of prop is function, only allow factory function
+  validator?(value: T): boolean
 }
 
 export type PropType<T> = PropConstructor<T> | PropConstructor<T>[]
 
-type PropConstructor<T = any> = { new (...args: any[]): T & object } | { (): T }
+type PropConstructor<T = any> = { new (): T & object } | { (): T }
```

Comment:
```suggestion
type PropConstructor<T = any> = { new (...args: never[]): T & object } | { (): T }
```

We put variant arguments here because to accept user defined class.
For some reason, it is not working in the latest TypeScript version but updating `any` with `never` fixes it.
I already made a PR to v2 repo for that. https://github.com/vuejs/vue/pull/10779

Reply:
@ktsn thanks